bundles/core/org.eclipse.smarthome.core.audio/src/main/java/org/eclipse/smarthome/core/audio/AudioManager.java

```diff
@@ -0,0 +1,460 @@
+                continue;
+            }
+
+            // As currentAudioFormat is concrete, use it
+            return currentAudioFormat;
+        }
+
+        // There's no concrete AudioFormat so we must create one
+        for (AudioFormat currentAudioFormat : audioFormats) {
+            // Define AudioFormat to return
+            AudioFormat format = currentAudioFormat;
+
+            // Not all Codecs and containers can be supported
+            if (null == format.getCodec()) {
+                continue;
+            }
+            if (null == format.getContainer()) {
+                continue;
+            }
+
+            // Prefer WAVE container
+            if (!format.getContainer().equals(AudioFormat.CONTAINER_WAVE)) {
+                continue;
+            }
+
+            // If required set BigEndian, BitDepth, BitRate, and Frequency to default values
+            if (null == format.isBigEndian()) {
+                format = new AudioFormat(format.getContainer(), format.getCodec(), new Boolean(true),
+                        format.getBitDepth(), format.getBitRate(), format.getFrequency());
+            }
+            if (null == format.getBitDepth() || null == format.getBitRate() || null == format.getFrequency()) {
+                // Define default values
+                int defaultBitDepth = 16;
+                long defaultFrequency = 16384;
+
+                // Obtain current values
+                Integer bitRate = format.getBitRate();
+                Long frequency = format.getFrequency();
+                Integer bitDepth = format.getBitDepth();
+
+                // These values must be interdependent (bitRate = bitDepth * frequency)
+                if (null == bitRate) {
+                    if (null == bitDepth) {
+                        bitDepth = new Integer(defaultBitDepth);
+                    }
+                    if (null == frequency) {
+                        frequency = new Long(defaultFrequency);
+                    }
+                    bitRate = new Integer(bitDepth.intValue() * frequency.intValue());
+                } else if (null == bitDepth) {
+                    if (null == frequency) {
+                        frequency = new Long(defaultFrequency);
+                    }
+                    bitDepth = new Integer(bitRate.intValue() / frequency.intValue());
+                } else if (null == frequency) {
+                    frequency = new Long(bitRate.longValue() / bitDepth.longValue());
+                }
+
+                format = new AudioFormat(format.getContainer(), format.getCodec(), format.isBigEndian(), bitDepth,
+                        bitRate, frequency);
+            }
+
+            // Return preferred AudioFormat
+            return format;
+        }
+
+        // Return null indicating failure
+        return null;
+    }
+
+    protected void addAudioSource(AudioSource audioSource) {
+        this.audioSources.put(audioSource.getId(), audioSource);
```

Comment:
This method is used by DS -- so it is called async.
The Map is used by other functions in this class, too.
There is no sync. used.

You can use a ConcurrentHashMap or add sync statements to prevent errors.